File: common/src/main/java/xyz/chunkstories/net/packets/PacketSoundSource.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.net.packets;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

import org.joml.Vector3d;
import org.joml.Vector3dc;

import xyz.chunkstories.api.client.net.ClientPacketsProcessor;
import xyz.chunkstories.api.exceptions.PacketProcessingException;
import xyz.chunkstories.api.net.PacketDestinator;
import xyz.chunkstories.api.net.PacketReceptionContext;
import xyz.chunkstories.api.net.PacketSender;
import xyz.chunkstories.api.net.PacketSendingContext;
import xyz.chunkstories.api.net.PacketWorld;
import xyz.chunkstories.api.sound.SoundSource;
import xyz.chunkstories.api.sound.SoundSource.Mode;
import xyz.chunkstories.api.world.World;
import xyz.chunkstories.sound.source.SoundSourceVirtual;

public class PacketSoundSource extends PacketWorld {
	public SoundSourceVirtual soundSourceToSend;

	public PacketSoundSource(World world) {
		super(world);
	}

	public PacketSoundSource(World world, SoundSourceVirtual soundSource) {
		this(world);
		this.soundSourceToSend = soundSource;
	}

	@Override
	public void send(PacketDestinator destinator, DataOutputStream out, PacketSendingContext ctx) throws IOException {
		out.writeUTF(soundSourceToSend.getName());
		out.writeLong(soundSourceToSend.getUuid());
		Vector3dc position = soundSourceToSend.getPosition();
		if (position != null) {
			out.writeBoolean(true);
			out.writeFloat((float) position.x());
			out.writeFloat((float) position.y());
			out.writeFloat((float) position.z());
		} else
			out.writeBoolean(false);
		// out.writeBoolean(soundSourceToSend.loop);
		// out.writeBoolean(soundSourceToSend.isAmbient);
		// out.writeBoolean(soundSourceToSend.buffered);
		out.writeByte(soundSourceToSend.getMode().ordinal());
		out.writeBoolean(soundSourceToSend.isDonePlaying());
		out.writeFloat(soundSourceToSend.getPitch());
		out.writeFloat(soundSourceToSend.getGain());
		out.writeFloat(soundSourceToSend.getAttenuationStart());
		out.writeFloat(soundSourceToSend.getAttenuationEnd());
	}

	@Override
	public void process(PacketSender sender, DataInputStream in, PacketReceptionContext processor)
			throws IOException, PacketProcessingException {
		String soundName = in.readUTF();
		long UUID = in.readLong();

		boolean hasPosition = in.readBoolean();
		Vector3dc position = null;
		if (hasPosition) {
			position = new Vector3d(in.readFloat(), in.readFloat(), in.readFloat());
		}

		// boolean loop = in.readBoolean();
		// boolean isAmbient = in.readBoolean();
		// boolean buffered = in.readBoolean();

		byte modeByte = in.readByte();
		Mode mode = Mode.values()[modeByte];

		boolean stopped = in.readBoolean();
		float pitch = in.readFloat();
		float gain = in.readFloat();
		float attenuationStart = in.readFloat();
		float attenuationEnd = in.readFloat();

		if (!(processor instanceof ClientPacketsProcessor))
			return;

		ClientPacketsProcessor cpe = (ClientPacketsProcessor) processor;

		SoundSource soundSource = cpe.getContext().getSoundManager().getSoundSourceByUUID(UUID);

		// ALSoundSource soundSource = (ALSoundSource)
		// Client.getInstance().getSoundManager().getSoundSourceByUUID(UUID);
		if (soundSource == null && stopped)
			return;

		if (soundSource == null) {

			soundSource = cpe.getContext().getSoundManager().replicateServerSoundSource(soundName, mode, position,
					pitch, gain, attenuationStart, attenuationEnd, UUID);
			return;
		}
		if (stopped) {
			soundSource.stop();
			return;
		}

		// Update the soundSource with all we can
		soundSource.setPosition(position);
		soundSource.setPitch(pitch);
		soundSource.setGain(gain);
		soundSource.setAttenuationStart(attenuationStart);
		soundSource.setAttenuationEnd(attenuationEnd);
	}

}

File: common/src/main/java/xyz/chunkstories/net/packets/PacketSoundSource.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//
package xyz.chunkstories.net.packets

import org.joml.Vector3d
import org.joml.Vector3dc
import xyz.chunkstories.api.client.Client
import xyz.chunkstories.api.net.PacketWorld
import xyz.chunkstories.api.player.Player
import xyz.chunkstories.api.sound.SoundSource
import xyz.chunkstories.api.world.World
import xyz.chunkstories.api.world.WorldSub
import xyz.chunkstories.sound.source.SoundSourceVirtual
import java.io.DataInputStream
import java.io.DataOutputStream

class PacketSoundSource(world: World) : PacketWorld(world) {
    var soundSourceToSend: SoundSourceVirtual? = null

    constructor(world: World, soundSource: SoundSourceVirtual) : this(world) {
        soundSourceToSend = soundSource
    }

    override fun send(dos: DataOutputStream) {
        dos.writeUTF(soundSourceToSend!!.soundName)
        dos.writeLong(soundSourceToSend!!.id)
        val position = soundSourceToSend!!.position
        if (position != null) {
            dos.writeBoolean(true)
            dos.writeFloat(position.x().toFloat())
            dos.writeFloat(position.y().toFloat())
            dos.writeFloat(position.z().toFloat())
        } else dos.writeBoolean(false)
        dos.writeByte(soundSourceToSend!!.mode.ordinal)
        dos.writeBoolean(soundSourceToSend!!.isDonePlaying)
        dos.writeFloat(soundSourceToSend!!.pitch)
        dos.writeFloat(soundSourceToSend!!.gain)
        dos.writeFloat(soundSourceToSend!!.attenuationStart)
        dos.writeFloat(soundSourceToSend!!.attenuationEnd)
    }

    override fun receive(dis: DataInputStream, player: Player?) {
        val soundName = dis.readUTF()
        val sourceId = dis.readLong()
        val hasPosition = dis.readBoolean()
        var position: Vector3dc? = null
        if (hasPosition) {
            position = Vector3d(dis.readFloat().toDouble(), dis.readFloat().toDouble(), dis.readFloat().toDouble())
        }
        val modeByte = dis.readByte()
        val mode = SoundSource.Mode.values()[modeByte.toInt()]
        val stopped = dis.readBoolean()
        val pitch = dis.readFloat()
        val gain = dis.readFloat()
        val attenuationStart = dis.readFloat()
        val attenuationEnd = dis.readFloat()

        val worldClient = world as? WorldSub ?: throw Exception("Not a client !")
        val client = worldClient.gameInstance as Client

        var soundSource: SoundSource? = client.soundManager.getSoundSource(sourceId)

        if (soundSource == null && stopped) return
        if (soundSource == null) {
            soundSource = client.soundManager.replicateServerSoundSource(soundName, mode, position!!, pitch, gain, attenuationStart, attenuationEnd, sourceId)
            return
        }
        if (stopped) {
            soundSource.stop()
            return
        }

        // Update the soundSource with all we can
        soundSource.position = position
        soundSource.pitch = pitch
        soundSource.gain = gain
        soundSource.attenuationStart = attenuationStart
        soundSource.attenuationEnd = attenuationEnd
    }
}